### ISSUE TITLE
Manage private DNS TXT record

### ISSUE THREAD

**Issue by carlos4ndre** [Cisco-Talos/clamav-faq#143]
Hi,

I'm setting up one private mirror following the steps described on [Serve CVD and CDIFF files from a local web server](https://www.clamav.net/documents/private-local-mirrors#serve-cvd-and-cdiff-files-from-a-local-web-server), and came across an issue when the private mirror is out of sync with the public mirror, worst case scenario for a time period no longer then the TTL of the TXT record `current.cvd.clamav.net`, due to DNS caching when checking for new updates.

The implications of this, is that for a high number of clients the load could be significant, as clients will not find the latest `.cdiff` file and will download the `daily.cvd` for each retry (file size is over 100MB).

Think this could be avoided if we controlled the information being published on the DNS TXT record, and that is actually possible if we own the DNS TXT record and reference it via the `DNSDatabaseInfo` setting.

However, I cannot find any information on how to populate this record. Is this approach discouraged? If not, could you provide more information on how to setup a DNS TXT record explaining its format (e.g., 0.103.0:59:25954:1602487740:0:63:49191:331) and potentially update the documentation?

Thanks!

**Comment by micahsnyder** [comment]
Hi @carlos4ndre apologies for the delay.

I asked around a bit and did some research of my own to try to answer your question.  Hosting your own DNS TXT record info is an option, but definitely places a larger burden on you.  If you want to do this, the record format is as follows:
`latest version:main cvd version:daily cvd version:record time:new security patch boolean:minimum functionaity level:safebrowsing cvd version:bytecode cvd version`

So for `0.103.0:59:25954:1602487740:0:63:49191:331` that's...
main.cvd version: 59
daily.cvd version: 25954
bytecode.cvd version: 25954
record time: 1602487740   (aka 10/12/2020 @ 7:29am (UTC))
new security patch boolean: 0  (false)
minimum functionaity level: 63   (see https://www.clamav.net/documents/functionality-levels-flevels for more details)
safebrowsing cvd version: 49191
bytecode cvd version: 331

HOWEVER, I don't think that maintaining your own DNS records is the way to go.  

Instead of hosting your own DNS TXT record server, the ideal solution would be to set `DNSDatabaseInfo no` so that freshclam will fail to use DNS for the version check and revert back to checking the version with an HTTP GET.  The freshclam clients will use an "If-Modified-Since" HTTP header and the response from the server will be a 304 "Not Modified" if they're up-to-date.

It would look something like this (in verbose mode):
```
* Connected to 0.0.0.0 (127.0.0.1) port 8000 (#0)
> GET /daily.cvd HTTP/1.1
Host: 0.0.0.0:8000
Range: bytes=0-511
User-Agent: ClamAV/0.103.0 (OS: Darwin, ARCH: x86_64, CPU: x86_64)
Accept: */*
If-Modified-Since: Mon, 26 Oct 2020 12:22:44 GMT
Connection: close

* HTTP 1.0, assume close after body
< HTTP/1.0 304 Not Modified
< Server: SimpleHTTP/0.6 Python/3.7.6
< Date: Mon, 26 Oct 2020 21:56:39 GMT
<
* Closing connection 0
daily database version up-to-date, according to HTTP response code from server.
```

UNFORTUNATELY I've uncovered a bug when investigating this.  It seems that the freshclam clients use the database published-date for the "If-Modified-Since" timestamp rather than the date the files were downloaded. Meanwhile, your server will use the "modified" timestamp for the local files to determine whether or not to send a 304 or send the file.  As a consequence, the server's modified time will certainly be older than our publish date, so it will send back the CVD header (512 bytes) and then client will check the publish date in the header to determine if it needs to request the entire CVD.  This is _still probably easier/better_ than hosting your own DNS TXT records as the version checks will download 512 bytes each time they check for an update rather than over 100MB when there is a version sync issue.  

If your database server supports HTTP "Range" requests, this will look something like this (verbose mode):

```
* Connected to 0.0.0.0 (127.0.0.1) port 8000 (#0)
> GET /daily.cvd HTTP/1.1
Host: 0.0.0.0:8000
Range: bytes=0-511
User-Agent: ClamAV/0.103.0 (OS: Darwin, ARCH: x86_64, CPU: x86_64)
Accept: */*
If-Modified-Since: Mon, 26 Oct 2020 12:22:44 GMT
Connection: close

* HTTP 1.0, assume close after body
< HTTP/1.0 206 Partial Content
< Server: SimpleHTTP/0.6 Python/3.7.6
< Date: Mon, 26 Oct 2020 23:29:16 GMT
< Content-type: application/octet-stream
< Accept-Ranges: bytes
< Content-Range: bytes 0-511/113917078
< Content-Length: 512
< Last-Modified: Mon, 26 Oct 2020 23:28:31 GMT
<
Time:    0.0s, ETA:    0.0s [========================>]       512B/512B
* Closing connection 0
OK
daily database version obtained using HTTP GET: 25969
daily.cvd database is up to date (version: 25969, sigs: 4335803, f-level: 63, builder: raynman)
fc_update_database: daily.cvd already up-to-date.
```

Would this solve your bandwidth concern?

The takeaway for me is that we should update the freshclam clients to use the databases' "modified" timestamps for our "If-Modified-Since" checks.  Specifically, [the `localTimestamp` variable here](https://github.com/Cisco-Talos/clamav-devel/blob/dev/0.104/libfreshclam/libfreshclam_internal.c#L1754) should be changed to use the file's "modified" timestamp in GMT / UTC time using something like this https://stackoverflow.com/questions/10446526/get-last-modified-time-of-file-in-linux 

I can put in a ticket for this bug fix, but we're swamped for 0.103.1 and 0.104 development right now so I'm not sure when we'll get to it.  If you or someone else wants to work on this, we love pull requests (https://github.com/Cisco-Talos/clamav-devel). 

Speaking of which, if you're interested in helping us improve the documentation here in https://github.com/Cisco-Talos/clamav-faq/, pull requests are very welcome. 

-Micah

**Comment by carlos4ndre** [comment]
Hi @micahsnyder,

Thanks for your detailed response, very insightful 👍 

That was my initial approach, disable the DNS version checks and download the CVD files whenever there's a new update.

The only issue with this approach was the less efficient use of bandwidth and scalability concerns with traffic as the number of clients increases. This could be solved somehow if there was some sort of metadata file with the current CVD versions that freshclam could fetch and then use to sort out the remaining CDIFF files. One example of this, would be the [index file](https://helm.sh/docs/topics/chart_repository/#the-index-file) stored on the Helm chart repository that clients use to discover the latest versions.

This could be an interesting feature, as an alternative to the DNS version checks when hosting a private mirror, and still serve CDIFF files.

So for now, I'm going to host the DNS record to optimize the bandwidth usage.

Thanks for you help!
